rfcs/2022-05-17-11532-chronicle-sink.md

```diff
@@ -0,0 +1,239 @@
+# RFC 11532 - 2022-05-17 - Google Chronicle Sink
+
+This RFC proposes a new sink that sends data into
+[Google Chronicle](https://cloud.google.com/chronicle/docs).
+Chronicle is a Security Information and Event Managent (SIEM) platform to store,
+aggregate and search security telemetry.
+
+## Context
+
+- https://github.com/vectordotdev/vector/issues/11532
+
+## Cross cutting concerns
+
+- The schema support required for the UDM events relies on
+  [this PR](https://github.com/vectordotdev/vector/pull/11743)
+  being merged.
+
+## Scope
+
+### In scope
+
+- A sink that sends data into Google Chronicle. This includes UDM events and
+  unstructured log events.
+
+### Out of scope
+
+- Adding any explicit knowledge of SIEM events as a separate event
+  type into Vector will not be considered.
+
+## Pain
+
+## Proposal
+
+### User Experience
+
+This adds a new sink to Vector to send data into Google Chronicle.
+
+There is a choice of two different types of events that can be sent to Chronicle.
+
+Which type to send will be configured with an `event_type` configuration option.
+
+```toml
+event_type = "unstructured"
+```
+
+or
+
+```toml
+event_type = "udm"
+```
+
+#### UDM (Unified Data Model) events
+
+https://cloud.google.com/chronicle/docs/reference/ingestion-api#udmevents
+https://cloud.google.com/chronicle/docs/unified-data-model/udm-usage
+
+UDM events are events that conform to a [strict schema](https://cloud.google.com/chronicle/docs/reference/udm-field-list).
+An example event:
+
+```json
+{
+  "metadata": {
+    "event_timestamp": "2019-10-23T04:00:00.000Z",
+    "event_type": "NETWORK_HTTP",
+    "product_name": "Acme Proxy",
+    "vendor_name": "Acme"
+  },
+  "network": {
+    "http": {
+      "method": "GET",
+      "response_code": 200
+    }
+  },
+  "principal": {
+    "hostname": "host0",
+    "ip": "10.1.2.3",
+    "port": 60000
+  },
+  "target": {
+    "hostname": "www.altostrat.com",
+    "ip": "198.51.100.68",
+    "port": 443,
+    "url": "www.altostrat.com/images/logo.png"
+  }
+}
+```
+
+Events can be batched and sent in a JSON object that also includes the
+`customer_id`:
+
+```json
+{
+"customer_id": "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c",
+"events": [...]
+}
+```
+
+UDM events can be sumbitted as either JSON or Proto3. We will submit using JSON
```

Comment:
What are the tradeoffs by adopting JSON? Are there performance tradeoffs?

Reply:
Protobufs will result in a lower network load. A quick search brings up [this comparison](https://nilsmagnus.github.io/post/proto-json-sizes/) tldr, savings could be 50%-80% smaller.

In terms of development complexity, with protobufs we would need to write a fair chunk of arduous code to convert between a log event and a protobuf object. With json it's a simple `serde_json::to_string(&log)` call.

I definitely think in the long term protobufs would be a higher quality solution, but suggest that it might be worth waiting to see if there is sufficient user demand before spending the extra development time.

Reply:
Thanks. Agree, probably useful context to add to the RFC.